src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -189,44 +185,72 @@ void ArchiveHeapWriter::ensure_buffer_space(size_t min_bytes) {
 void ArchiveHeapWriter::copy_roots_to_buffer(GrowableArrayCHeap<oop, mtClassShared>* roots) {
   Klass* k = Universe::objectArrayKlass(); // already relocated to point to archived klass
   int length = roots->length();
-  _heap_roots_word_size = objArrayOopDesc::object_size(length);
-  size_t byte_size = _heap_roots_word_size * HeapWordSize;
-  if (byte_size >= MIN_GC_REGION_ALIGNMENT) {
-    log_error(cds, heap)("roots array is too large. Please reduce the number of classes");
-    vm_exit(1);
-  }
-
-  maybe_fill_gc_region_gap(byte_size);
-
-  size_t new_used = _buffer_used + byte_size;
-  ensure_buffer_space(new_used);
 
-  HeapWord* mem = offset_to_buffered_address<HeapWord*>(_buffer_used);
-  memset(mem, 0, byte_size);
-  {
-    // This is copied from MemAllocator::finish
-    oopDesc::set_mark(mem, markWord::prototype());
-    oopDesc::release_set_klass(mem, k);
-  }
-  {
-    // This is copied from ObjArrayAllocator::initialize
-    arrayOopDesc::set_length(mem, length);
-  }
-
-  objArrayOop arrayOop = objArrayOop(cast_to_oop(mem));
-  for (int i = 0; i < length; i++) {
-    // Do not use arrayOop->obj_at_put(i, o) as arrayOop is outside of the real heap!
-    oop o = roots->at(i);
-    if (UseCompressedOops) {
-      * arrayOop->obj_at_addr<narrowOop>(i) = CompressedOops::encode(o);
-    } else {
-      * arrayOop->obj_at_addr<oop>(i) = o;
+  // Depending on the number of classes we are archiving, a single roots array may be
+  // larger than MIN_GC_REGION_ALIGNMENT. Roots are allocated first in the buffer, which
+  // allows us to chop the large array into a series of "segments". Current layout
+  // starts with zero or more segments exactly fitting MIN_GC_REGION_ALIGNMENT, and end
+  // with a single segment that may be smaller than MIN_GC_REGION_ALIGNMENT.
+  // This is simple and efficient. We do not need filler objects anywhere between the segments,
+  // or immediately after the last segment. This allows starting the object dump immediately
+  // after the roots.
+
+  assert((_buffer_used % MIN_GC_REGION_ALIGNMENT) == 0,
+         "Pre-condition: Roots start at aligned boundary: " SIZE_FORMAT, _buffer_used);
+
+  int max_elem_count = ((MIN_GC_REGION_ALIGNMENT - arrayOopDesc::header_size_in_bytes()) / heapOopSize);
+  assert(objArrayOopDesc::object_size(max_elem_count)*HeapWordSize == MIN_GC_REGION_ALIGNMENT,
+         "Should match exactly");
+
+  HeapRoots heap_roots(_buffer_used,
+                       roots->length(),
+                       MIN_GC_REGION_ALIGNMENT,
+                       max_elem_count);
+
+  int segment_beg = 0;
+  while (segment_beg < length) {
+    // Guess the size of next segment.
+    int elem_count = MIN2(max_elem_count, length - segment_beg);
+    size_t bytes_size = objArrayOopDesc::object_size(elem_count) * HeapWordSize;
+
+    size_t oop_offset = _buffer_used;
+    _buffer_used = oop_offset + bytes_size;
```

Comment:
Assert that `_buffer_used` is region aligned.
Assert that `bytes_size` is the same as MIN_GC_REGION_ALIGNMENT

Reply:
Hm, neither of these work, because the trailing segment might not fill the entire region (by design). We really only care that segment starts are aligned at region size, which current assert for `oop_offset` checks.